### PR TITLE
Desktop: Go To Anything by body

### LINKED ISSUE #2683: Search in note body using GotoAnything
Currently GotoAnything allows searching within note titles, notebook titles and tags.

We want to add this feature: When pressing the `/` key, followed by a query, GotoAnything should search within the note body.

All code should be in GotoAnything.jsx

Forum thread: https://discourse.joplinapp.org/t/what-shortcut-for-gotoanything-to-search-in-note-body/2148

### PR DESCRIPTION
**Pull request by anjulalk**
Closes #2683 

Using `/` prefix followed by a search query now shows list of notes which matches the query in their bodies. I had to add a new state variable `highlightKeywords` to prevent body search keywords getting highlighted in the titles.

I've also edited the help text. Hope that's not an issue for translation?

### CODE REVIEW

**Comment by laurent22** [comment]
Thanks for the pull request @anjulalk. Could you go a step further and display fragments of the note body below the title, with highlighted keywords?

For example, if the title is "Welcome to Joplin" with body "Joplin is a free, open source note taking and to-do application, which helps you write and organise your notes, and synchronise them between your devices.", and you search for "note". It would display this:

> Welcome to Joplin (note title)
> ... open source **note** taking and ... organise your **note**s, and synchronise ... (note body)
> Welcome (notebook)

So title and notebook stay the same, but when searching within the content, you would display a line in between showing the highlighted results. You don't need to display all the results, maybe just the first 2 or 3.

**Comment by anjulalk** [comment]
That'd be a great addition indeed. I've came up with a way to implement this, but I don't know if it's the best way. Therefore please give your feedback.

The results from the search engine currently don't include note bodies. So we can conditionally include the note bodies. (Only if user is searching body. Otherwise note bodies are not included.) Then we can use regular expressions to find locations which matches the query, and then display their surrounding content as you've mentioned. 

The problem is it's not very efficient. Is there a better way? Can we get the matching locations from SQL?

**Comment by laurent22** [comment]
> The problem is it's not very efficient. Is there a better way? Can we get the matching locations from SQL?

Indeed I'd be worried there's potentially a lot of content to be returned. Could you for example only load the body from the first 100 notes from GotoAnything.jsx, and display this? Then beyond 50 notes, you simply display "..." in the results to show that it's not a bug, but we just aren't loading this.

You can use `Note.byIds(noteIds, { fields: ["id", "body"]})` to load only the first 100 notes, based on the IDs you got from the search engine.

**Comment by anjulalk** [comment]
Great! I'm working on it. I'll give an update soon.

**Comment by laurent22** [comment]
Any update?

**Comment by anjulalk** [comment]
Sorry, it took longer than I expected. I've managed to implement the functionality but I think it may not be perfect. Therefore please give feedback.

Screenshot at 2020-03-15 22-27-05

For each keyword match, it finds beginning and end indices of the word containing the keyword. (Actually it finds the ending of the word which is 15 characters further than the actual keyword, to get some more words) For example, searching for 'not' gives the string `note taking and to-do`. It saves these indices.

There can be multiple indices overlapping, for example, `note taking and to-do` for `note` keyword, and `taking and to-do application` for `taking` keyword. Then it merges these indices into one: `note taking and to-do application`.

Finally it strips these strings from the body, and join them with ` ... `

**Comment by anjulalk** [comment]
Example: fragments for `note` and `taking` keywords are merged
image

File: ElectronClient/plugins/GotoAnything.jsx [hunk 1]
```diff
@@ -165,9 +186,55 @@ class Dialog extends React.PureComponent {
...
+								stack.push(indices[j]);
+							} else if (top[1] < indices[j][1]) {
+								top[1] = indices[j][1];
+								stack.pop();
+								stack.push(top);
+							}
+						}
+
+						fragments = stack.map(f => body.slice(f[0], f[1])).join(' ... ');
+					}
```
Comment (laurent22) [inline ElectronClient/plugins/GotoAnything.jsx]: Could you refactor this into a function to make it clearer what it does?

File: ReactNativeClient/lib/string-utils.js [hunk 2]
```diff
@@ -264,6 +264,15 @@ function substrWithEllipsis(s, start, length) {
...
 }
 
+function closestWordIndices(s, match, offset) {
+	const begin = s.slice(0, (match.index >>> 0) + 1).search(/\S+$/);
+	let end = s.slice(match.index + match[0].length + offset).search(/\s/);
+	if (end < 0) end = s.length;
+	else end += match.index + match[0].length + offset;
+
+	return [begin, end];
+}
```
Comment (laurent22) [inline ReactNativeClient/lib/string-utils.js]: Please add a comment to describe:
- What this function does
- What are the parameters
- What is the output

Comment (anjulalk) [reply ReactNativeClient/lib/string-utils.js]: I've changed this function to nextWhitespaceIndex(s, begin) which seeks the next whitespace character from `begin` index in string `s` and return its index. It's much more simpler. Let me know if I should make any changes.

File: ElectronClient/plugins/GotoAnything.jsx (hunk 1 above)
Comment (anjulalk) [reply ElectronClient/plugins/GotoAnything.jsx]: I've refactored the merge code into a function and added comments to make it more understandable.

**Comment by anjulalk** [comment]
Thanks a lot for the review. I've made the requested changes and added more comments. 
The code should be more cleaner and simpler now. Please have a look.

Please let me know of any required changes. I'll make them right away!

**Comment by laurent22** [comment]
Thanks @anjulalk, code looks good now and your comments make it clear how the loop to build up the text fragments works.

Could you just add unit tests for mergeOverlappingIntervals and nextWhitespaceIndex please? I guess you can move mergeOverlappingIntervals to ArrayUtils as it will make it easier to test it. You'll find the relevant test units in tests/ArrayUtils.js and tests/StringUtils.js

**Comment by anjulalk** [comment]
I've moved the `mergeOverlappingIntervals()` to `ArrayUtils.js` file, and added unit tests for `mergeOverlappingIntervals` and `nextWhitespaceIndex()` functions. 
Please have a look and give feedback. Thanks!

File: CliClient/tests/ArrayUtils.js [hunk 3]
```diff
@@ -49,4 +49,29 @@ describe('ArrayUtils', function() {
 		expect(ArrayUtils.contentEquals(['b'], ['a', 'b'])).toBe(false);
 	}));
 
+	it('should merge overlapping intervals', asyncTest(async () => {
+		const testCases = [
```
Comment (laurent22) [inline CliClient/tests/ArrayUtils.js]: Could you add a test case for the empty array please?

File: CliClient/tests/StringUtils.js [hunk 4]
```diff
@@ -41,4 +41,18 @@ describe('StringUtils', function() {
 		}
 	}));
 
+	it('should find the next whitespace character', asyncTest(async () => {
+		const str = 'Joplin is a free, open source\n note taking and *to-do* application';
+		const testCases = [[0, 6], [12, 17], [23, 29], [48, 54]];
```
Comment (laurent22) [inline CliClient/tests/StringUtils.js]: Likewise, please could you add two test cases:
- for str = ""
- and for when there's no whitespace after the specified index

**Comment by anjulalk** [comment]
Apologies, I should've checked those edge cases before. I've added them and fixed the error caused by `mergeOverlappingIntervals()` when providing an empty array. Please have a look.

**Comment by laurent22** [comment]
Thanks for the update @anjulalk, it's good to go now.
